### PR TITLE
Add diff logic between two AMTs

### PR DESCRIPTION
**Pull request by iand**
This is an update to some existing work by @arajasek that brings it up to date with the refactored node types and support for different bit widths.

### CODE REVIEW

File: diff.go [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+		return removeAll(ctx, prevBs, prev, prevBitWidth, prevHeight, offset)
+	}
+
+	if prevHeight == 0 && curHeight == 0 {
+		return diffLeaves(prev, cur, offset)
+	}
+
+	changes := make([]*Change, 0)
+
+	if curHeight > prevHeight {
```
Comment (Stebalien) [inline diff.go:75]: Can we dedup this code with the code below?

File: diff.go [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+		subCount := nodesForHeight(curBitWidth, curHeight)
+
+		// TODO: cur.links may be nil
```
Comment (Stebalien) [inline diff.go:78]: This should be fine, right? It just means it's empty.

**Review by Stebalien — APPROVED** [review]
This code looks correct and well-tested, but it would be nice to dedup it a bit if possible.

File: diff.go (hunk 2 above)
Comment (iand) [reply diff.go:78]: Yep, removed that TODO

File: diff.go (hunk 1 above)
Comment (iand) [reply diff.go]: I looked and although they are structurally identical the numbers and roles of the parameters needed gives a shared function a pretty complex signature and doesn't save much code.

I simplified the function signatures as part of this but I don't see a good way to reduce it further at the moment.

File: diff.go [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+
+	if cur == nil {
+		return removeAll(ctx, prevCtx, prev, offset)
+	}
+
+	if prevCtx.height == 0 && curCtx.height == 0 {
+		return diffLeaves(prev, cur, offset)
+	}
+
+	changes := make([]*Change, 0)
```
Comment (Stebalien) [inline diff.go:96]: nit: specifying the length isn't useful. I'd just leave this as:

```
var changes []*Change
```

or 

```
changes := make([]*Change)
```

**Review by Stebalien — APPROVED** [review]
This _looks_ correct. I'm still pretty sure we can dedup some of this code with some careful helper functions, but this is a good place to start.

**Comment by iand** [comment]
ping @Stebalien  (i can't merge here)

**Pull request merged**